cirq-google/cirq_google/calibration/xeb_wrapper.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import multiprocessing
+
+import cirq
+import cirq.experiments.random_quantum_circuit_generation as rqcg
+import cirq.experiments.xeb_fitting as xebf
+import cirq.experiments.xeb_sampling as xebsamp
+from cirq_google.calibration.phased_fsim import (
+    XEBPhasedFSimCalibrationRequest,
+    PhasedFSimCalibrationResult,
+    PhasedFSimCharacterization, XEBPhasedFSimCalibrationOptions,
+)
+
+
+def run_calibration(
+        calibration: XEBPhasedFSimCalibrationRequest,
```

Comment:
I was hoping that for this calibration you'd create something like `LocalXEBPhasedFSimCalibrationRequest` which indicates that XEB should be done locally.

This way we could keep the single [run_calibrations](https://github.com/quantumlib/Cirq/blob/master/cirq-google/cirq_google/calibration/workflow.py#L602) method which triggers calibrations. The method could recognize the local requests and delegate them to your wrapper. The method takes a sequence of calibrations  so one issue would be if user mixes local and non-local calibrations. We could disallow for such a mix and raise an exception or filter out local and non-local calibrations.

What I would like to preserve is an ability of decoupling sampler/engine from calibration requests: user should write their code against `cirq.google.run_calibrations` and depending on passed sampler and different type of requests, the desired functionality could be realized.

There is a todo for me here as well, I see that `run_calibrations` take `engine: Union[Engine, PhasedFSimEngineSimulator]`. I think it will be better if it is just `engine: Union[Engine, Sampler]`, and if `EngineSampler` is passed, the engine object could be extracted from there. This will simplify our tutorials as well. This is related to #3788 but I'm not sure if this is the tame to make it so public, we need to test this feature more.

Reply:
Sounds good.

The dispatch behavior of `cirq.google.run_calibrations` is getting really complicated. 

1. normal requests, PhasedFSimEngineSimulator [calls special engine.get_calibrations]
2. normal requests, Engine [calls engine.run_calibration]
3. local requests, Engine [construct enginesampler, calls wrapper run_local_xeb_calibration]
4. local requests, sampler [calls wrapper run_local_xeb_calibration]

You could potentially mix-and-match (2) and (3), but that gets tricky when max_layers_per_request > 1 and they get batched together. I think for now, we should enforce only one calibration type per call to run_calibrations.

I'll put together an example with the above four conditions, but let me know if you have any thoughts.

Reply:
I think we are getting somewhere with this. I'm fine with restricting single-type calibrations per request. In fact. I'm starting to miss more and more a grouping object which we could use to group calibrations together.

The only case that I'm hesitant is number (1) and local sampler. Should it also call engine.get_calibrations? Calling sampler and doing real simulation is useful but I'm not sure if this is the right abstraction level. On the other hand, engine.get_calibrations() doesn't support fidelities that could be obtained automatically by doing simulations. Maybe boolean option to run_calibrations for that?

Please take a look at #4054, I have simplified `run_calibrations` independently; we can merge with this PR.